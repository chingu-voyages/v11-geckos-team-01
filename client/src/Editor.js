import PropTypes from 'prop-types';

require('codemirror/lib/codemirror.css');
require('codemirror/theme/material.css');
require('codemirror/theme/neat.css');
require('codemirror/mode/xml/xml.js');
require('codemirror/mode/javascript/javascript.js');
require('codemirror/addon/lint/lint');
require('codemirror/addon/lint/json-lint');

// eslint-disable-next-line import/first
import './Editor.css';
// eslint-disable-next-line import/first
import { formatJSONfromString } from './Utils';

// eslint-disable-next-line import/first
import initialValue from './initial.js';
// JSON.parse(formatJSONfromString(initialValue))

// eslint-disable-next-line import/first
import React from 'react';

class Editor extends React.Component {
  constructor(props) {
    super(props);
    this.textAreaNode = null;
    this.codeMirrorInstance = null;
    this.codeMirror = null;
    this.options = {
      mode: props.mode ? props.mode : 'javascript',
      theme: 'material',
      viewportMargin: props.viewPortMargin ? props.viewPortMargin : 10,
      lineNumbers: true
    };
    this.state = {
      lastTemplateId: null,
      defaultValue: initialValue
    }
  }
  getCodeMirrorInstance() {
    return require('codemirror');
  }
  componentWillReceiveProps(props) {
    const { lastTemplateId } = this.state
    if (props.newTemplateId !== lastTemplateId) {
      this.setState({ lastTemplateId: props.newTemplateId })
      try {
        this.codeMirror
          .getDoc()
          .setValue(JSON.stringify(props.defaultValue, null, 2));
      } catch (error) {
        console.error(error)
        return false;
      }
    }
  }
  componentDidMount() {
    this.setState({
      defaultValue: this.props.defaultValue
    })
    const codeMirrorInstance = this.getCodeMirrorInstance();
    this.codeMirror = codeMirrorInstance.fromTextArea(
      this.textAreaNode,
      this.options
    );
    //
    // Saves the editor content whenever a change happens
    //
    const doc = this.codeMirror.getDoc();
    const str = formatJSONfromString(doc.getValue());
    console.log(str)
    const val = JSON.parse(str);
    this.props.onChange(val);

    this.codeMirror.on('change', (doc, change) => {
      try {
        // const str = formatJSONfromString(doc.getValue());
        const str = JSON.parse(doc.getValue())
        // const val = JSON.parse(str);
        this.props.onChange(str);
      } catch (error) {
        console.error(error);
      }
    });
  }
  render() {
    return (
      <textarea
        ref={(ref) => (this.textAreaNode = ref)}
        autoFocus={true}
        defaultValue={this.state.defaultValue}
      ></textarea>
    );
  }
}

Editor.propTypes = {
  codeMirrorInstance: PropTypes.func,
  onChange: PropTypes.func,
  options: PropTypes.object,
  readOnly: PropTypes.bool,
  defaultValue: PropTypes.array,
  newTemplateId: PropTypes.string,
  preserveScrollPosition: PropTypes.bool
};

export default Editor;
