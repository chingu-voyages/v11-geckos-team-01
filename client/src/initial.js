const initial = `[{
  'repeat(50, 100)': {
    _id: '{{guid}}',
    company: [
      {
        'repeat(3, 5)': {
	  	  _id: '{{guid}}',
          name: '{{companyName}}',
          value: '{{amount}}'
        }
      }
    ],
    people: [{
      'repeat(3, 5)': {
        firstName: '{{firstName}}',
        lastName: '{{lastName}}',
		_id: '{{guid}}',
        facts: [{
          'repeat(3, 5)': {
            _id: '{{guid}}',
            desc: '{{words}}'
          }
       }]
      }
    }]
  }
}]`
export default initial
