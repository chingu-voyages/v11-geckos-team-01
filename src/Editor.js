

require('codemirror/lib/codemirror.css')
require('codemirror/theme/material.css')
require('codemirror/theme/neat.css')
require('codemirror/mode/xml/xml.js')
require('codemirror/mode/javascript/javascript.js')
require('codemirror/addon/lint/lint');
require('codemirror/addon/lint/json-lint');

// eslint-disable-next-line import/first
import './Editor.css'
// eslint-disable-next-line import/first
import { formatJSONfromString } from './Utils'

// eslint-disable-next-line import/first
import PropTypes from 'prop-types'

// eslint-disable-next-line import/first
import React from 'react'

class Editor extends React.Component {
  constructor(props) {
    super(props)
    console.log(props.mode)
    this.textAreaNode = null
    this.codeMirrorInstance = null
    this.codeMirror = null
    this.options = {
      mode: props.mode ? props.mode : 'javascript',
      theme: 'material',
      viewportMargin: props.viewPortMargin ? props.viewPortMargin : 10,
      lineNumbers: true
    }
    this.state = {
      // defaultValue: `\nfunction Foo () {\n  return "Bar"\n}\n`
      defaultValue: null
    }
  }
  getCodeMirrorInstance () {
		return require('codemirror')
  }
  componentWillReceiveProps (props) {
    if (props.readOnly) {
      console.log(props)
      try {
        this.codeMirror.getDoc().setValue(
          JSON.stringify(JSON.parse(props.defaultValue), null, 2)
        )
      } catch (error) {
        return false
      }
    }
  }
  componentDidMount () {
    const codeMirrorInstance = this.getCodeMirrorInstance()
    this.codeMirror = codeMirrorInstance.fromTextArea(this.textAreaNode, this.options)
    //
    // Saves the editor content whenever a change happens
    //
    console.log(this.props.defaultValue)
    console.log(this.props.readOnly)
    this.setState({ defaultValue: this.props.defaultValue })
    if (!this.props.readOnly) {
      this.codeMirror.on('change', (doc, change) => {
        try {
          const str = formatJSONfromString(doc.getValue())
          this.props.onChange(JSON.parse(str))
        } catch (error) {
          console.error(error)
        }
      })
    }
    // this.codeMirror.setSize('100%', '100%')
  }
  render() {
    return (
      <textarea
        ref={(ref) => this.textAreaNode = ref}
        autoFocus={true}
        readOnly={this.props.readOnly}
        defaultValue={this.props.defaultValue}
      ></textarea>
    )
  }
}

Editor.propTypes = {
  autoFocus: PropTypes.bool,
  className: PropTypes.any,
  codeMirrorInstance: PropTypes.func,
  // intialValue: PropTypes.string,
  name: PropTypes.string,
  onChange: PropTypes.func,
  onCursorActivity: PropTypes.func,
  onFocusChange: PropTypes.func,
  onScroll: PropTypes.func,
  options: PropTypes.object,
  path: PropTypes.string,
  value: PropTypes.string,
  readOnly: PropTypes.bool,
  defaultValue: PropTypes.string,
  preserveScrollPosition: PropTypes.bool,
}

export default Editor
