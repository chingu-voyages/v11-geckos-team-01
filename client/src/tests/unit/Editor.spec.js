
import React from 'react';
import ReactDOM from 'react-dom';

import { shallow, mount } from 'enzyme';

import Editor from '../../components/Editor.js'

import codeMirrorSetup from '../helpers/codeMirrorSetup.js';

import initial from '../../initial.json'

console.log(typeof initial);

const mountFunction = (options = {}) => {

  let { onChange, defaultValue } = options

  if (!onChange) {
    onChange = () => {}
  }

  return mount(
    <Editor
      onChange={onChange}
      viewPortMargin={Infinity}
      newTemplateId={''}
      readOnly={false}
    />
  );
}

beforeEach(codeMirrorSetup)

it('renders without crashing', () => {
  mountFunction()
});

it('should invoke onChange callback on initial mount', () => {
  const onChange = jest.fn();

  const wrapper = mountFunction({ onChange })

  expect(onChange).toHaveBeenCalled();
})

it('should display an intial value that matches a snapshot', () => {
  const wrapper = mountFunction();

  const el = wrapper.find('#editor-instance');

  expect(el.text()).toMatchSnapshot();
})
