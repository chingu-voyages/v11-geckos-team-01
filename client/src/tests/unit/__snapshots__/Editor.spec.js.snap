// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display an initial value upon rendering 1`] = `
"{
  \\"data\\": {
    \\"id\\": \\"{{random.uuid}}\\",
    \\"isDeleted\\": true,
    \\"percentage\\": 0,
    \\"createdOn\\": \\"2019-10-24T20:08:45.889Z\\",
    \\"createdBy\\": \\"{{name.lastName}}\\",
    \\"updatedOn\\": \\"2019-10-24T20:08:45.889Z\\",
    \\"updatedBy\\": \\"{{name.lastName}}\\"
  },
  \\"version\\": \\"{{random.number}}\\",
  \\"statusCode\\": \\"{{random.number}}\\",
  \\"requestId\\": \\"{{random.uuid}}\\",
  \\"timestamp\\": 0
}"
`;

exports[`should display an intial value that matches a snapshot 1`] = `
"{
  \\"data\\": {
    \\"id\\": \\"{{random.uuid}}\\",
    \\"isDeleted\\": true,
    \\"percentage\\": 0,
    \\"createdOn\\": \\"2019-10-24T20:08:45.889Z\\",
    \\"createdBy\\": \\"{{name.lastName}}\\",
    \\"updatedOn\\": \\"2019-10-24T20:08:45.889Z\\",
    \\"updatedBy\\": \\"{{name.lastName}}\\"
  },
  \\"version\\": \\"{{random.number}}\\",
  \\"statusCode\\": \\"{{random.number}}\\",
  \\"requestId\\": \\"{{random.uuid}}\\",
  \\"timestamp\\": 0
}"
`;
